### PR TITLE
Testing for Default Inference for Device Type

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69802
* #70209
* #69051
* #69050
* #71184
* __->__ #69052
* #69049

Differential Revision: [D33555888](https://our.internmc.facebook.com/intern/diff/D33555888)

### CODE REVIEW

File: test/jit/test_device_analysis.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        self.assertEqual(len(graph_inputs), len(example_devices))
+        for graph_i, device_i in zip(graph_inputs, example_devices):
+            if device_i is not None:
+                graph_i.setType(graph_i.type().with_device(device_i))
+
+        torch._C._jit_pass_propagate_device(graph)
+
+    def assert_device_equal(self, fn, in_devices, expected_device):
+        graph = torch.jit.script(fn).graph
+        print(graph)
```
Comment (Krovatkin) [inline test/jit/test_device_analysis.py:43]: nit: stray print?

File: test/jit/test_device_analysis.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        def add_self(x):
+            return x + x
+
+        graph = torch.jit.script(add_self).graph
+        graph_input = next(graph.inputs())
+        graph_input.setType(graph_input.type().with_device(self.cpu))
+        # self.prop_device_on_graph(graph, [self.cpu])
+        self.assertEqual(graph_input.type().device(), self.cpu)
+
+    # TODO: I have fixed all the bugs with the input types being applied. However
```
Comment (Krovatkin) [inline test/jit/test_device_analysis.py:60]: could you please elaborate, does `setType(device)` really work? Most tests in this PR rely on it to work.

File: test/jit/test_device_analysis.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        def set_device(x):
+            return x.to("cpu")
+
+        for in_device in self.device_types:
+            self.assert_device_equal(set_device, [in_device], self.cpu)
+
+    def test_device_arg(self):
+        # Test that no device gets propagated when arg is passed in
+        def set_device(x, device_name: torch.device):
+            return x.to(device=device_name)
```
Comment (Krovatkin) [inline test/jit/test_device_analysis.py:85]: can we also test `cpu()`, `cuda()` these are pretty popular too or do they get desugared to `.to(cpu)` , `to(cuda)`?

File: test/jit/test_device_analysis.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            self.assert_device_equal(fn, [in_device], in_device)
+
+    def test_set_dtype(self):
+        def set_device(x):
+            return x.to("cpu")
+
+        for in_device in self.device_types:
+            self.assert_device_equal(set_device, [in_device], self.cpu)
+
+    def test_device_arg(self):
```
Comment (Krovatkin) [inline test/jit/test_device_analysis.py:82]: a side note, for both dtype and device tests, we should probably have some bigger tests with mutltiple operations maybe even mobilenet.

**Review by Krovatkin — COMMENTED** [review]
a few qq

File: test/jit/test_device_analysis.py (hunk 3 above)
Comment (Gamrix) [reply test/jit/test_device_analysis.py:85]: Those are in a subsequent diff.

File: test/jit/test_device_analysis.py (hunk 4 above)
Comment (Gamrix) [reply test/jit/test_device_analysis.py:82]: Sounds good, but out of scope for this diff. Will put something for that in a later diff in the this (or another) stack.

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: test/jit/test_device_analysis.py (hunk 4 above)
Comment (Gamrix) [reply test/jit/test_device_analysis.py:97]: Added testing for MobileNet on this diff later in the stack, as it needs other parts of device type analysis.
Adding custom device type change rules #69051

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33555888).

**Pull request closed without merging**